### PR TITLE
add function of identifying the line ranges of the errant code snippets

### PR DESCRIPTION
**Pull request by mannaandpoem**
**Features**
add function of identifying the line ranges of the errant code snippets

### CODE REVIEW

**Review by geekan — CHANGES_REQUESTED** [review]
```
isort....................................................................Failed
- hook id: isort
- files were modified by this hook

Fixing /home/runner/work/MetaGPT/MetaGPT/swe_bench/inference/run_api.py

ruff.....................................................................Passed
black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted swe_bench/inference/run_agent.py
reformatted swe_bench/gitagent.py
```

File: swe_bench/inference/run_agent.py [hunk 1]
```diff
@@ -30,8 +30,12 @@ def _prepare(inputs):
     cleaned_user_message = re.sub("<patch>.*?</patch>", "", user_message, flags=re.DOTALL)
 
     issues = re.findall("<issue>(.*?)</issue>", user_message, flags=re.DOTALL)
+    issues = [re.sub(r"#{3,4} Versions.*?(?=#{3,4}|\Z)", "", issues[0], flags=re.DOTALL)]
+    traceback = re.findall(r"#{3,4} Actual Results.*", issues[0], flags=re.DOTALL)
+    issues = traceback if traceback else issues
+    code = re.findall("<code>(.*?)</code>", user_message, flags=re.DOTALL)
```
Comment (geekan) [inline swe_bench/inference/run_agent.py:36]: Can you explain the corresponding effects of these changes in the issue body?

Comment (mannaandpoem) [reply swe_bench/inference/run_agent.py:72]: The purpose is：
1. Extract the actual result from the issue
2. Extract all code from message

File: data/load_dataset.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+# @Author  : stellahong (stellahong@fuzhi.ai)
+# @Desc    :
+from pathlib import Path
+
+import numpy as np
+from datasets import load_dataset, load_from_disk
+
+from data.inference.const import SCIKIT_LEARN_IDS
```
[the remark this answers is not in the document]
Comment (mannaandpoem) [reply data/load_dataset.py:9]: OK, let me modify it

File: benchmark/swe_bench/utils/ast_parser.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                node_full_name = current.name + "." + node_full_name
+        return self.file_name + "." + node_full_name
+
+    def get_node_full_name_by_import(self, node):
+        if hasattr(node, "end_lineno"):  # Python 3.8+
+            # 获取从开始行号到结束行号的所有行内容
+            import_statement_lines = self.source_code.splitlines()[node.lineno - 1 : node.end_lineno]
+            complete_statement = "".join([line.strip() for line in import_statement_lines])
+        else:  # 在Python 3.8之前的版本中，没有end_lineno
+            # 只能获取开始的行
```
Comment (stellaHSR) [inline benchmark/swe_bench/utils/ast_parser.py:37]: Please convert comments into English.

File: benchmark/swe_bench/utils/jaccard_retriever.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from typing import Union
+
+from transformers import AutoTokenizer
+
+
+def jaccard_retriever(code, candidates: list, similarity: str = "jaccard"):
+    tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-350M-multi", cache_dir="cache")
```
Comment (stellaHSR) [inline benchmark/swe_bench/utils/jaccard_retriever.py:7]: add docstring and input parameters description.

File: benchmark/swe_bench/utils/repo_controller.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+from pathlib import Path
+from typing import Optional
+
+import git
+
+from metagpt.logs import logger
+
+
+class RepoController:
```
Comment (stellaHSR) [inline benchmark/swe_bench/utils/repo_controller.py:9]: Most of the functions have been implemented in utils/repo_utils.

**Comment by mannaandpoem** [comment]
TODO: Improve the method of segmenting relevant code based on the retrieved function name by using retrieve

**Pull request closed without merging**